### PR TITLE
Add helpers for setting/getting data in `node.run_state['sensu']`, make ssl data bags optional

### PR DESCRIPTION
**Pull request by cwjohnston**
These changes are proposed to address concerns around required data bags for SSL (see #260, #306, #325). The intent here is to keep the default behavior but at the same time provide a means for wrapper cookbooks to implement their own method for injecting credentials into the Chef run without adding dependencies in this cookbook.

I'd like to add unit tests for these helpers but I'm not sure how to implement them. Thoughts and constructive criticism are welcome!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -44,12 +44,31 @@ This cookbook makes no attempt to manage the versions of its package dependencie
...
+}
+```
+
+All of the above values are expected to be strings comprised of PEM-formatted credentials with escaped line endings. See `test/integration/data_bags/sensu/ssl.json` for a more literal example.
+
+If the attempt to load SSL credentials from a data bag fails, the cookbook will log a warning but proceed with the rest of the Chef run anyway, on the assumption that credentials will be inserted into the Chef "run state" (i.e. `node.run_state['sensu']['ssl']`) in the same format using the `Sensu::ChefRunState` helper methods, `set_sensu_run_state` and `get_sensu_run_state`.
+
+Please see the [documentation for the run state helper methods](#helper-modules-and-methods) for more information.
+
+This cookbook comes with a tool to generate the certificates and datam bag items. If the integrity of the certificates is ever compromised, you must regenerate and redeploy them.
```
Comment (luckymike) [inline README.md]: datam?

File: README.md [hunk 2]
```diff
@@ -233,6 +252,43 @@ sensu_snippet "irc" do
 end
 ```
 
+## Helper modules and methods
+
+### Run State Helpers
+
+The `Sensu::ChefRunState` module provides helper methods which populate `node.run_state['sensu']` with arbitrary key/value pairs. This provides a means for wrapper cookbooks to populate the `node.run_state` with data required by the cookbook, e.g. SSL credentials, without cookbook itself enforcing source for that data.
+
+n.b. The `node.run_state` is not persisted locally nor on a Chef server. Data stored here exists only for the durration of the Chef run.
```
Comment (luckymike) [inline README.md]: s/durration/duration

**Comment by luckymike** [comment]
This looks sound to me and seems like it improves usability in wrapper cookbooks. 

For unit testing, can we use the helpers to set and get a known arbitrary value? It seems like that is as much as unit testing can reasonably cover, and the formatting and ingestion of such a value should be part of an integration test in the wrapper.

File: README.md (hunk 1 above)
Comment (agoddard) [inline README.md]: I LIKE IT!

**Comment by cwjohnston** [comment]
Added integration tests with secrets specified statically in the `sensu-test::run_state_helpers` recipe.

**Comment by agoddard** [comment]
:+1:

**Comment by pghalliday** [comment]
This solves the data bag problem for me.

Feels a little over engineered though, i didn't realise that setting the run state was so complicated it needed to be wrapped ;)

Also the purist in me wants to remove the data bag stuff entirely - but i can understand the desire for backward compatibility

**Comment by luckymike** [comment]
Sensu Enterprise isn't starting for me correctly with just this branch, but if I rebase develop (e.g. the merged state), things work, so I'm considering this :+1: 

I think there's certainly an argument to be made for removing the data bag stuff so that is entirely a wrapper function, but I think that should be a future consideration.

**Pull request merged**